exporters/autoexport/registry.go

```diff
@@ -0,0 +1,138 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package autoexport // import "go.opentelemetry.io/contrib/exporters/autoexport"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/sdk/trace"
+)
+
+const (
+	otelExporterOtlpProtoEnvKey = "OTEL_EXPORTER_OTLP_PROTOCOL"
+)
+
+// registry maintains a map of exporter names to SpanExporter
+// implementations that is safe for concurrent use by multiple goroutines
+// without additional locking or coordination.
+type registry struct {
+	mu    sync.Mutex
+	names map[string]trace.SpanExporter
+}
+
+var (
+	// envRegistry is the index of all supported exporter types
+	// and their mapping to a SpanExporter.
+	envRegistry = &registry{
+		names: map[string]trace.SpanExporter{},
+	}
+
+	// errUnknownExpoter is returned when an unknown exporter name is used in
+	// the OTEL_*_EXPORTER environment variables.
+	errUnknownExpoter = errors.New("unknown exporter")
+
+	// errInvalidOtlpProtocol is returned when an invalid protocol is used in
+	// the OTEL_EXPORTER_OTLP_PROTOCOL environment variable.
+	errInvalidOtlpProtocol = errors.New("invalid OTLP protocol - should be one of ['grpc', 'http/protobuf']")
+)
+
+// load returns the value stored in the registry index for a key, or nil if no
+// value is present. The ok result indicates whether value was found in the
+// index.
+func (r *registry) load(key string) (p trace.SpanExporter, ok bool) {
+	r.mu.Lock()
+	p, ok = r.names[key]
+	r.mu.Unlock()
+	return p, ok
+}
+
+var errDupReg = errors.New("duplicate registration")
+
+// store sets the value for a key if is not already in the registry. errDupReg
+// is returned if the registry already contains key.
+func (r *registry) store(key string, value trace.SpanExporter) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if r.names == nil {
+		r.names = map[string]trace.SpanExporter{key: value}
+		return nil
+	}
+	if _, ok := r.names[key]; ok {
+		return fmt.Errorf("%w: %q", errDupReg, key)
+	}
+	r.names[key] = value
+	return nil
+}
+
+// drop removes key from the registry if it exists, otherwise nothing.
+func (r *registry) drop(key string) {
+	r.mu.Lock()
+	delete(r.names, key)
+	r.mu.Unlock()
+}
+
+// RegisterSpanExporter sets the SpanExporter e to be used when the
+// OTEL_TRACES_EXPORTERS environment variable contains the exporter name. This
+// will panic if name has already been registered.
+func RegisterSpanExporter(name string, e trace.SpanExporter) {
```

Comment:
I know Start is only an OTLP exporter concept, but do you see any issues that might arise from packages registering a "started" client that isn't used?

eg. If I register a jaeger, an otlp, and a zipkin with the intent of the operator choosing via environment variables, will having misconfigured other exporters sitting in the registry cause problems?

Reply:
I don't think so, started exporters aren't actually wired up into a pipeline as part of this package and an exporter should only act on data to export. Started exporters may hold resources that would be idle and unused.

Reply:
Sorry I never got back to this @dashpole.

I think a factory would be useful to defer the span exporter creation, though it would mean we'd need to maintain a list of factories alongside the tranceExporter instances too. 

Maybe should can look at doing that as a follow-up optimsation, I don't think it's needed to start with.

Reply:
> Exporters are not expected to be concurrent safe and if we return the same exporter to all requests it may be they are registered with different span processors. I think for this reason the factor makes more sense here as we want to return a unique instance.

Maybe we can solve problem of uniqueness with adding additional properties to `envRegistry map` key? I would just avoid another level of indirection (factory in this case) until it's really necessary